app/src/main/java/com/quran/labs/androidquran/widgets/HighlightingImageView.java

```diff
@@ -130,17 +152,178 @@ public void setNightMode(boolean isNightMode, int textBrightness) {
+    }
+
+    final String ayahTransition = previousSurahAyah+"->"+currentSurahAyah;
+
+    // yes we make copies, because normalizing the bounds will change them
+    List<AyahBounds> previousAyahBoundsList = new ArrayList<>(startingBounds);
+    List<AyahBounds> currentAyahBoundsList = new ArrayList<>(coordinatesData.get(currentSurahAyah));
+
+    highlights.clear();
+    highlights.add(ayahTransition);
+
+    // TODO: reason whether this TypeEvaluator should be anonymous or not
+    animator = ValueAnimator.ofObject(new TypeEvaluator() {
+
+      private void normalizeAyahBoundsList(List<AyahBounds> start, List<AyahBounds> end) {
+        // this function takes two unequal length lists and normalizes them
+
+        int startSize = start.size();
+        int endSize = end.size();
+        int minSize = Math.min(startSize, endSize);
+        int maxSize = Math.max(startSize, endSize);
+        List<AyahBounds> minList = startSize < endSize? start : end;
+        int diff = maxSize - minSize;
+
+        RectF rectToBeDivided = minList.get(minSize-1).getBounds();
+        float oLeft = rectToBeDivided.left;
+        float oRight = rectToBeDivided.right;
+        float oTop = rectToBeDivided.top;
+        float oBottom = rectToBeDivided.bottom;
+        minList.remove(minSize-1);
+        float part = (oRight-oLeft) /(diff+1);
+        for(int i=0; i<(diff+1); ++i) {
+          float left = oLeft + part*i;
+          float right = left + part;
+          RectF rect = new RectF(left, oTop, right, oBottom);
+          AyahBounds ayahBounds = new AyahBounds(0, 0, rect);
+          minList.add(ayahBounds);
+        }
+      }
+
+      @Override
+      public Object evaluate(float fraction, Object startObject, Object endObject) {
+        List<AyahBounds> start = (List<AyahBounds>)startObject;
+        List<AyahBounds> end = (List<AyahBounds>)endObject;
+
+        if(start.size() != end.size()) {
+          normalizeAyahBoundsList(start, end);
+        }
+
+        int size = start.size();
+
+        //TODO: instead of creating result every time, modify in-place
+        List<AyahBounds> result = new ArrayList<>(size);
+
+        for(int i=0; i<size; ++i) {
+          RectF startValue = start.get(i).getBounds();
+          RectF endValue = end.get(i).getBounds();
+          float left = startValue.left + (endValue.left - startValue.left) * fraction;
+          float top = startValue.top + (endValue.top - startValue.top) * fraction;
+          float right = startValue.right + (endValue.right - startValue.right) * fraction;
+          float bottom = startValue.bottom + (endValue.bottom - startValue.bottom) * fraction;
+          AyahBounds intermediateBounds = new AyahBounds(0,0, new RectF(left, top, right, bottom));
+          result.add(intermediateBounds);
+        }
+        return result;
+      }
+    }, previousAyahBoundsList, currentAyahBoundsList);
+
+    // TODO: where should this constant live?
```

Comment:
if you move a lot of this out to new classes, let's move it there

Reply:
Have moved this to HighlightAnimationConfig